modules/pricer/src/main/java/com/opengamma/strata/pricer/fx/ImpliedTrinomialTreeFxSingleBarrierOptionProductPricer.java

```diff
@@ -0,0 +1,470 @@
+   * <p>
+   * The trinomial tree is first calibrated to Black volatilities, 
+   * then the price is computed based on the calibrated tree.
+   * 
+   * @param option  the option product
+   * @param ratesProvider  the rates provider
+   * @param volatilityProvider  the Black volatility provider
+   * @return the currency exposure
+   */
+  public MultiCurrencyAmount currencyExposure(
+      ResolvedFxSingleBarrierOption option,
+      RatesProvider ratesProvider,
+      BlackVolatilityFxProvider volatilityProvider) {
+
+    RecombiningTrinomialTreeData data = calibrateTrinomialTree(option, ratesProvider, volatilityProvider);
+    return currencyExposure(option, ratesProvider, volatilityProvider, data);
+  }
+
+  /**
+   * Calculates the currency exposure of the FX barrier option product.
+   * <p>
+   * This assumes the tree is already calibrated and the tree data is stored as {@code RecombiningTrinomialTreeData}.
+   * The tree data should be consistent with the pricer and other inputs, see {@link #validateData}.
+   * 
+   * @param option  the option product
+   * @param ratesProvider  the rates provider
+   * @param volatilityProvider  the Black volatility provider
+   * @param treeData  the trinomial tree data
+   * @return the currency exposure
+   */
+  public MultiCurrencyAmount currencyExposure(
+      ResolvedFxSingleBarrierOption option,
+      RatesProvider ratesProvider,
+      BlackVolatilityFxProvider volatilityProvider,
+      RecombiningTrinomialTreeData treeData) {
+
+    ResolvedFxVanillaOption underlyingOption = option.getUnderlyingOption();
+    ValueDerivatives priceDerivatives = priceDerivatives(option, ratesProvider, volatilityProvider, treeData);
+    double price = priceDerivatives.getValue();
+    double delta = priceDerivatives.getDerivative(0);
+    CurrencyPair currencyPair = underlyingOption.getUnderlying().getCurrencyPair();
+    double todayFx = ratesProvider.fxRate(currencyPair);
+    double signedNotional = signedNotional(underlyingOption);
+    CurrencyAmount domestic = CurrencyAmount.of(currencyPair.getCounter(), (price - delta * todayFx) * signedNotional);
+    CurrencyAmount foreign = CurrencyAmount.of(currencyPair.getBase(), delta * signedNotional);
+    return MultiCurrencyAmount.of(domestic, foreign);
+  }
+
+  //-------------------------------------------------------------------------
+  /**
+   * Calibrate trinomial tree to Black volatilities. 
+   * <p>
+   * The calibration procedure is in principle independent of the option to price. 
+   * However, {@code ResolvedFxSingleBarrierOption} is plugged in to ensure that the grid points properly cover 
+   * the lifetime of the target option. 
+   * 
+   * @param option  the option
+   * @param ratesProvider  the rates provider
+   * @param volatilityProvider  the Black volatility provider
+   * @return the trinomial tree data
+   */
+  public RecombiningTrinomialTreeData calibrateTrinomialTree(
```

Comment:
This calibration is generic to a given expiry and FX. The bulk of this method should probably be in calculator in 'pricer.fx'. It can also be used for vanilla FX options.